ludwig/serve.py

```diff
@@ -0,0 +1,156 @@
+
+
+def server(
+    model,
+):
+    app = FastAPI()
+
+    input_features = {
+        f['name'] for f in model.model_definition['input_features']
+    }
+
+    @app.post('/predict')
+    async def predict(request: Request):
+        form = await request.form()
+        files, entry = convert_input(form)
+
+        try:
+            if (entry.keys() & input_features) != input_features:
+                return JSONResponse(ALL_FEATURES_PRESENT_ERROR,
+                                    status_code=400)
+            try:
+                resp = model.predict(data_dict=[entry]).to_dict('records')[0]
+                return JSONResponse(resp)
+            except Exception as e:
+                logger.error("Error: {}".format(str(e)))
+                return JSONResponse(COULD_NOT_RUN_INFERENCE_ERROR,
+                                    status_code=500)
+        finally:
+            for f in files:
+                os.remove(f.name)
+    return app
+
+
+def convert_input(form):
+    '''
+    Returns a new input and a list of files to be cleaned up
+    '''
+    new_input = {}
+    files = []
+    for k, v in form.multi_items():
+        if type(v) == UploadFile:
+            # Convert UploadFile to a NamedTemporaryFile to ensure it's on the disk
+            suffix = os.path.splitext(v.filename)[1]
+            named_file = tempfile.NamedTemporaryFile(
+                delete=False, suffix=suffix)
+            files.append(named_file)
+            named_file.write(v.file.read())
+            named_file.close()
+            new_input[k] = named_file.name
+        else:
+            new_input[k] = v
+
+    return (files, new_input)
+
+
+def cli(sys_argv):
+    parser = argparse.ArgumentParser(
+        description='This script serves a pretrained model',
+        prog='ludwig serve',
+        usage='%(prog)s [options]'
+    )
+
+    # ----------------
+    # Model parameters
+    # ----------------
+    parser.add_argument(
+        '-m',
+        '--model_path',
+        help='model to load',
+        required=True
+    )
+
+    parser.add_argument(
+        '-l',
+        '--logging_level',
+        default='info',
+        help='the level of logging to use',
+        choices=['critical', 'error', 'warning', 'info', 'debug', 'notset']
+    )
+
+    # ----------------
+    # Server parameters
+    # ----------------
+    parser.add_argument(
+        '-p',
+        '--port',
+        help='port for server (default: 8000)',
+        default=8000,
+        type=int,
+    )
+
+    parser.add_argument(
+        '-H',
+        '--host',
+        help='host for server (default: 0.0.0.0)',
+        default='0.0.0.0'
+    )
+
+    args = parser.parse_args(sys_argv)
+
+    logging.getLogger('ludwig').setLevel(
+        logging_level_registry[args.logging_level]
+    )
+
+    model = LudwigModel.load(args.model_path)
+    app = server(model)
+    uvicorn.run(app, host=args.host, port=args.port)
```

Comment:
Maybe these 3 lines can be packaged in a single `serve_model(model_path, host, port)` function, so that If someone wants to start a server programmatically, or maybe starting more than one server programmatically on different ports they can do it.